istioctl/cmd/add-to-mesh.go

```diff
@@ -188,15 +268,231 @@ func findDeploymentsForSvc(client kubernetes.Interface, ns, name string) ([]apps
+		Resource: "serviceentries",
+	}
+	_, err = dynamicClient.Resource(serviceEntryGVR).Namespace(ns).Get(resourceName(opts.Name), metav1.GetOptions{})
+	if err == nil {
+		return fmt.Errorf("service entry %q already exists, skip", resourceName(opts.Name))
+	}
+
+	if err = generateServiceEntry(u, opts); err != nil {
+		return err
+	}
+	generateK8sService(s, opts)
+	if err = createServiceEntry(dynamicClient, ns, u, opts.Name, writer); err != nil {
+		return err
+	}
+	return createK8sService(client, ns, s, writer)
+}
+
+func generateServiceEntry(u *unstructured.Unstructured, o *vmServiceOpts) error {
+	if o == nil {
+		return fmt.Errorf("empty vm service options")
+	}
+	ports := []*v1alpha3.Port{}
+	for _, p := range o.PortList {
+		ports = append(ports, &v1alpha3.Port{
+			Number:   uint32(p.Port),
+			Protocol: string(p.Protocol),
+			Name:     p.Name,
+		})
+	}
+	eps := []*v1alpha3.ServiceEntry_Endpoint{}
+	for _, ip := range o.IP {
+		eps = append(eps, &v1alpha3.ServiceEntry_Endpoint{
+			Address: ip,
+			Labels:  o.Labels,
+		})
+	}
+	host := fmt.Sprintf("%v.%v.svc.cluster.local", o.Name, o.Namespace)
+	spec := &v1alpha3.ServiceEntry{
+		Hosts:      []string{host},
+		Ports:      ports,
+		Endpoints:  eps,
+		Resolution: v1alpha3.ServiceEntry_STATIC,
+	}
+	u.Object["spec"] = spec
+	return nil
+}
+
+func resourceName(hostShortName string) string {
+	return fmt.Sprintf("mesh-expansion-%v", hostShortName)
+}
+
+func generateK8sService(s *corev1.Service, o *vmServiceOpts) {
+	ports := []corev1.ServicePort{}
+	for _, p := range o.PortList {
+		ports = append(ports, corev1.ServicePort{
+			Name: strings.ToLower(p.Name),
+			Port: int32(p.Port),
+		})
+	}
+
+	spec := corev1.ServiceSpec{
+		Ports: ports,
+	}
+	s.Spec = spec
+}
+
+func convertToMap(s []string) map[string]string {
+	out := make(map[string]string, len(s))
+	for _, l := range s {
+		k, v := splitEqual(l)
+		out[k] = v
+	}
+	return out
+}
+
+// splitEqual splits key=value string into key,value. if no = is found
+// the whole string is the key and value is empty.
+func splitEqual(str string) (string, string) {
+	idx := strings.Index(str, "=")
+	var k string
+	var v string
+	if idx >= 0 {
+		k = str[:idx]
+		v = str[idx+1:]
+	} else {
+		k = str
+	}
+	return k, v
+}
+
+// createK8sService creates k8s service object for external services in order for DNS query and cluster VIP.
+func createK8sService(client kubernetes.Interface, ns string, svc *corev1.Service, writer io.Writer) error {
+	if svc == nil {
+		return fmt.Errorf("failed to create vm service")
+	}
+	if _, err := client.CoreV1().Services(ns).Create(svc); err != nil {
+		return fmt.Errorf("failed to create kuberenetes service %v", err)
+	}
+	if _, err := client.CoreV1().Services(ns).UpdateStatus(svc); err != nil {
+		return fmt.Errorf("failed to create kuberenetes service %v", err)
+	}
+	sName := strings.Join([]string{svc.Name, svc.Namespace}, ".")
+	fmt.Fprintf(writer, "Service %q has been created in the Istio service mesh"+
```

Comment:
We should be explicit that this is a Kube service to avoid confusion.

Reply:
fmt.Fprintf(writer, "Kubernetes Service %q has been created in the Istio